drivers: spi: Initial support for GD32 spi

cameled opened the pull request:
Add initail support for GigaDevice GD32 SPI controller.

More GD32 support can be found in #38657.

[EDIT]
TODO-List:
- [x] pinctrl for spi, use @gmarull GD32 pinctrl
- [x] spi_flash sample available for gd32vf103v_eval board
- [x] software CS, use @gmarull GD32 gpio
- [x] clock control, for max-frequency check.
~~ASYNC support~~(close this feature)

gmarull commented:
@cameled please, rebase. pinctrl is now available, so SPI driver can make use of it.

Commit:
dts: arm: introduce gd32 spi interface

Add gd32 spi initial support.

Signed-off-by: HaiLong Yang <cameledyang@pm.me>

Commit:
drivers: spi: introduce gd32 spi driver

Add a simple gd32 spi driver.

Signed-off-by: HaiLong Yang <cameledyang@pm.me>

cameled commented:
> please, rebase. pinctrl is now available, so SPI driver can make use of it.

Thanks @gmarull, Looks pinctrl work fine for SPI. I have successfully run the serial flash sample.

gmarull commented on boards/arm/gd32f450i_eval/gd32f450i_eval-pinctrl.dtsi:
```diff
@@ -11,4 +11,12 @@
+
+	spi5_default: spi5_default {
+		pins1 {
+			pinmux = <SPI5_SCK_PG13>,  <SPI5_NSS_PG8>,   \
+				 <SPI5_MOSI_PG14>, <SPI5_MISO_PG12>, \
+				 <SPI5_IO2_PG10>,  <SPI5_IO3_PG11>;
```
nit: `\` not needed:
```
			pinmux = <SPI5_SCK_PG13>, <SPI5_NSS_PG8>,
				 <SPI5_MOSI_PG14>, <SPI5_MISO_PG12>,
				 <SPI5_IO2_PG10>, <SPI5_IO3_PG11>;
```

are IO2/IO3 required? (only needed for quad mode, as far as I know). Depending on the SPI speed you may need to set `slew-rate`.

cameled replied:
> nit: `\` not needed:

Thanks for you advice, removed.

> are IO2/IO3 required? (only needed for quad mode, as far as I know). Depending on the SPI speed you may need to set `slew-rate`.

IO2/IO3 are not required, no quad spi driver for it now, remove it.

gmarull commented on boards/arm/gd32f450i_eval/gd32f450i_eval-pinctrl.dtsi:
```diff
@@ -11,4 +11,12 @@
 	};
+
+	spi5_default: spi5_default {
+		pins1 {
```
nit: `group1`

cameled replied:
done.

nandojve reviewed (changes requested):
nits

cameled commented on drivers/spi/Kconfig.gd32:
```diff
@@ -0,0 +1,19 @@
+	bool "Gigadevice GD32 SPI controller driver"
+	depends on SOC_FAMILY_GD32
+	select USE_GD32_SPI
+	help
+	  Enables Gigadevice GD32 SPI controller support.
+
+
+if SPI_GD32
```
Thanks!

cameled commented:
With @gmarull gpio_gd32 driver, spi cs control can select the device now.

cameled commented:
Ready to review, spi frequency select will implement at the next version.

gmarull commented on dts/bindings/spi/gd,gd32-spi.yaml:
```diff
@@ -0,0 +1,20 @@
+
+include: [spi-controller.yaml, pinctrl-device.yaml]
+
+properties:
```
suggestion: make `pinctrl-0` and `pinctrl-names` required.

cameled replied:
done.

gmarull commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,251 @@
+	uint32_t stat = SPI_STAT(spi);
+
+	if (stat & SPI_GD32_ERR_MASK) {
+		LOG_WRN("spi%u error status detected, err = %u",
```
`LOG_ERR`

cameled replied:
done.

gmarull commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,251 @@
+	}
+
+	if (SPI_OP_MODE_GET(config->operation) == SPI_OP_MODE_SLAVE) {
+		LOG_ERR("Slave mode not support");
```
`supported`

cameled replied:
done.

gmarull commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,251 @@
+		spi_parameter.frame_size = SPI_FRAMESIZE_16BIT;
+	}
+
+	if (config->cs) {
```
`!= NULL`

cameled replied:
done.

gmarull commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,251 @@
+		spi_parameter.clock_polarity_phase |= SPI_CTL0_CKPH;
+	}
+
+	/* Default spi clock prescaler to PCLK/32 */
+	spi_parameter.prescale = SPI_PSC_32;
```
nit: why always 32? could that be a DT parameter?

nandojve replied:
+1

cameled replied:
That's okay, I will add a DTS property to hold it.

cameled replied:
Since clock control is available at #41855, I will try to use it.

gmarull commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,251 @@
+	spi_context_lock(&data->ctx, false, NULL, config);
+
+	ret = spi_gd32_configure(dev, config);
+	if (ret) {
```
`ret < 0`

cameled replied:
done.

gmarull commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,251 @@
+
+	do {
+		ret = spi_gd32_frame_exchange(data, cfg->reg);
+	} while (!ret && spi_gd32_transfer_ongoing(data));
```
`(ret == 0) && ...`

cameled replied:
Adjust to 
```
	do {
		ret = spi_gd32_frame_exchange(dev);
		if (ret < 0) {
			break;
		}
	} while (spi_gd32_transfer_ongoing(data));
```

gmarull commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,251 @@
+		.rcu_periph_clock = DT_INST_PROP(idx, rcu_periph_clock),	\
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),			\
+	};									\
+	DEVICE_DT_INST_DEFINE(idx, &spi_gd32_init,				\
+			      NULL,						\
+			      &spi_gd32_data_##idx,				\
+			      &spi_gd32_config_##idx,				\
+			      POST_KERNEL,					\
+			      CONFIG_KERNEL_INIT_PRIORITY_DEVICE,		\
+			      &spi_gd32_driver_api);
```
```
	DEVICE_DT_INST_DEFINE(idx, &spi_gd32_init, NULL, &spi_gd32_data_##idx,	\
			      &spi_gd32_config_##idx, POST_KERNEL,		\
			      CONFIG_SPI_INIT_PRIORITY, &spi_gd32_driver_api);
```

cameled replied:
done.

gmarull commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,251 @@
+	/* Default spi clock prescaler to PCLK/32 */
+	spi_parameter.prescale = SPI_PSC_32;
+
+	spi_init(cfg->reg, &spi_parameter);
```
non-blocking: consider using lower level HAL register access (it's confusing to understand what is HAL vs Zephyr code here)

cameled replied:
That's a good ideal, I will try to use the raw register operator to implement it.

cameled replied:
done.

gmarull commented on boards/arm/gd32f450i_eval/gd32f450i_eval-pinctrl.dtsi:
```diff
@@ -17,4 +17,12 @@
+	spi5_default: spi5_default {
+		group1 {
+			pinmux = <SPI5_SCK_PG13>,  <SPI5_MOSI_PG14>,
+				 <SPI5_MISO_PG12>, <SPI5_IO2_PG10>,
+				 <SPI5_IO3_PG11>;
```
do IO2/IO3 need to be setup as AF even if they're not used? or are they?

gmarull reviewed (changes requested):
Thanks for this contribution, please see my comments.

gmarull commented on samples/boards/gd32/spi_gd25q16b/README.rst:
```diff
@@ -0,0 +1,29 @@
+.. _gd32_spi_gd25q16b:
```
note: you need to implement a flash driver instead (same as EEPROM for I2C)

cameled replied:
use spi_flash sample instead.
done.

Commit:
boards: add gd35q16 flash to gd32f450i_eval and gd32vf103v_eval boards

Add gd25q16 as "jedec,spi-nor" compatible flash to gd32f450i_eval and
gd32vf103v_eval boards.

Signed-off-by: HaiLong Yang <cameledyang@pm.me>

Commit:
samples: spi_flash: add gd32f450i_eval and gd32vf103v_eval boards

This add gd32f450i_eval and gd32vf103v_eval boards to spi_flash sample.

Signed-off-by: HaiLong Yang <cameledyang@pm.me>

cameled commented:
Enable the clock control feature, will sync with #41855.

cameled commented:
Thanks @gmarull,  gd32 clock-control ``on`` API works fine for gd32-spi driver, test with gd32f450i_eval board on spi_flash sample.
The next step is ``get_rate`` API confirmation, I will check it tomorrow.

cameled commented:
> Thanks @gmarull, gd32 clock-control `on` API works fine for gd32-spi driver, test with gd32f450i_eval board on spi_flash sample. The next step is `get_rate` API confirmation, I will check it tomorrow.

`get_rate` API work fine on spi-gd32 driver.

nandojve commented:
Hi folks,

After a long period of inactivity I think I'll have a little of free time to continue review.
Just to clarify, Is this just waiting clock-control?

cameled commented:
> After a long period of inactivity I think I'll have a little of free time to continue review. Just to clarify, Is this just waiting clock-control?

Yes, the SPI frequency select relay on clock-control.
 
> BTW, it will require fix CI errors.

Sure, It should be fixed. I working on another task now, May be fixed in the end of this week.

nandojve commented on dts/riscv/gigadevice/gd32vf103.dtsi:
```diff
@@ -123,6 +123,18 @@
 			label = "I2C_0";
 		};
 
+		spi0: spi@40013000 {
```
I think is missing SPI2/3

cameled replied:
Refer from [GD32 MCUs selection guide](http://gd32mcu.com/data/documents/qitaziliao/GD32_eries_of_MCUs_Selection_Guide(Jan_2022).pdf), gd32vf103 may have 1 or 3 spi interface. This only handle the basic case.

nandojve commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,276 @@
+		return -ENOTSUP;
+	}
+
+	/* Before configure, disable spi first. */
```
You can drop this comment.

cameled replied:
dropped.

nandojve commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,276 @@
+	/* Before configure, disable spi first. */
+	SPI_CTL0(cfg->reg) &= ~SPI_CTL0_SPIEN;
+
+	/* Default to master mode. */
```
You can drop this comment.

cameled replied:
dropped.

nandojve commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,276 @@
+
+	/* Default to master mode. */
+	SPI_CTL0(cfg->reg) |= SPI_MASTER;
+	/* Default to full duplex transfer mode. */
```
You can drop this comment.

cameled replied:
dropped.

nandojve commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,276 @@
+		SPI_CTL0(cfg->reg) |= SPI_CTL0_SWNSSEN;
+	} else {
+		/*
+		 * For signle master env,
```
nit: single

cameled replied:
Thanks! I'm start to using code spell checker , may help to avoid this miss spell issue.

nandojve commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,276 @@
+		SPI_CTL1(cfg->reg) |= SPI_CTL1_NSSDRV;
+	}
+
+	/* Reset to MSB first mode. */
```
You can drop this comment.

nandojve commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,276 @@
+		goto error;
+	}
+
+	/* Enable this spi peripheral. */
```
You can drop this comment.

nandojve commented on drivers/spi/spi_gd32.c:
```diff
@@ -0,0 +1,276 @@
+
+	spi_context_cs_control(&data->ctx, false);
+
+	/* Disable this spi peripheral. */
```
You can drop this comment.

gmarull commented on boards/arm/gd32f450i_eval/gd32f450i_eval-pinctrl.dtsi:
```diff
@@ -30,4 +30,13 @@
+		group1 {
+			pinmux = <SPI5_SCK_PG13>, <SPI5_MOSI_PG14>,
+				 <SPI5_MISO_PG12>,
+				 /* Use pinmux to pullup pg10 and pg11. */
+				 <SPI5_IO2_PG10>, <SPI5_IO3_PG11>;
```
nit: pull-up is not being enabled, does this mean SPI peripheral keeps these lines pulled-up when selected here?

cameled replied:
Yes.

If PG10 and PG11 configured as IO2 and IO3, it default behavior is output high level.

Better solution is use the WP# and HOLD# pins instead, but there doesn't have related logic in spi_nor flash driver.

nandojve reviewed (dismissed):
LGTM!

Thank you @cameled !

cameled commented:
@gmarull @nandojve Could you please re-approve this, I forget to update the commit message. Thanks!

MaureenHelm reviewed (commented):
Why is this DNM?
